src/OrasProject.Oras/Registry/Remote/Repository.cs

```diff
@@ -371,6 +397,205 @@ internal Reference ParseReferenceFromContentReference(string reference)
+                           .ConfigureAwait(false))
+        {
+            // If Referrers API is supported, then it would return referrers continuously
+            // otherwise, this line of code is not executed
+            // and the ReferrerState would be set to false in the method ReferrersByApi.
+            yield return referrer;
+        }
+        
+        if (ReferrersState == Referrers.ReferrersState.NotSupported)
+        {
+            // fall back to tag schema to retrieve referrers
+            await foreach (var referrer in ReferrersByTagSchema(descriptor, artifactType, cancellationToken)
+                               .ConfigureAwait(false))
+            {
+                yield return referrer;
+            }
+            
+            yield break;
+        }
+        
+        // Set ReferrersState to true as ReferrersState is not NotSupported/Unknown
+        SetReferrersState(true);
+    }
+
+    /// <summary>
+    /// ReferrersByApi retrieves a collection of referrers asynchronously based on the given descriptor and optional artifact type.
+    /// </summary>
+    /// <param name="descriptor"></param>
+    /// <param name="artifactType"></param>
+    /// <param name="cancellationToken"></param>
+    /// <returns></returns>
+    /// <exception cref="ResponseException"></exception>
+    /// <exception cref="NotSupportedException"></exception>
+    /// <exception cref="InvalidResponseException"></exception>
+    internal async IAsyncEnumerable<Descriptor> ReferrersByApi(Descriptor descriptor, string? artifactType,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        var reference = Options.Reference.Clone();
+        reference.ContentReference = descriptor.Digest;
+        var nextPageUrl = new UriFactory(reference).BuildReferrersUrl(artifactType);
+        
+        while (nextPageUrl != null)
+        {
+            // If ReferrerListPageSize is greater than 0, modify the URL to include the page size query parameter
+            if (ReferrerListPageSize > 0)
+            {
+                var uriBuilder = new UriBuilder(nextPageUrl);
+                var query = HttpUtility.ParseQueryString(uriBuilder.Query);
+                query.Add("n", ReferrerListPageSize.ToString());
+                uriBuilder.Query = query.ToString();
+                nextPageUrl = uriBuilder.Uri;
+            }
+
+            using var response = await _opts.HttpClient.GetAsync(nextPageUrl, cancellationToken).ConfigureAwait(false);
+            switch (response.StatusCode)
+            {
+                case HttpStatusCode.OK:
+                    // If the status code is OK, continue processing the response
+                    break;
+                case HttpStatusCode.NotFound:
+                    // If the status code is NotFound, handle as an error, possibly a non-existent repository
+                    var err = (ResponseException)await response.ParseErrorResponseAsync(cancellationToken)
```

Comment:
nit:
Since this is an exception, how about calling it `ex`?
```suggestion
                    var ex = await response.ParseErrorResponseAsync(cancellationToken)
```